third_party/proton/csrc/lib/Driver/GPU/Roctracer.cpp

```diff
@@ -0,0 +1,104 @@
+#include "Driver/GPU/Roctracer.h"
+#include "Driver/Dispatch.h"
+
+namespace proton {
+
+namespace roctracer {
+
+struct ExternLibRoctracer : public ExternLibBase {
+  using RetType = roctracer_status_t;
+  static constexpr const char *name = "libroctracer64.so";
+  static constexpr RetType success = ROCTRACER_STATUS_SUCCESS;
+  static void *lib;
+};
+
+void *ExternLibRoctracer::lib = nullptr;
+
+DEFINE_DISPATCH(ExternLibRoctracer, setProperties, roctracer_set_properties,
+                roctracer_domain_t, void *)
+
+DEFINE_DISPATCH(ExternLibRoctracer, getTimestamp, roctracer_get_timestamp,
+                roctracer_timestamp_t *)
+
+void start() {
+  typedef void (*roctracer_start_t)();
+  static roctracer_start_t func = nullptr;
+  Dispatch<ExternLibRoctracer>::init(ExternLibRoctracer::name,
+                                     &ExternLibRoctracer::lib);
+  if (func == nullptr)
+    func = reinterpret_cast<roctracer_start_t>(
+        dlsym(ExternLibRoctracer::lib, "roctracer_start"));
+  if (func)
+    func();
+}
+
+void stop() {
+  typedef void (*roctracer_stop_t)();
+  static roctracer_stop_t func = nullptr;
+  Dispatch<ExternLibRoctracer>::init(ExternLibRoctracer::name,
+                                     &ExternLibRoctracer::lib);
+  if (func == nullptr)
+    func = reinterpret_cast<roctracer_stop_t>(
+        dlsym(ExternLibRoctracer::lib, "roctracer_stop"));
+  if (func)
+    func();
+}
+
+char *getOpString(uint32_t domain, uint32_t op, uint32_t kind) {
+  typedef char *(*roctracer_op_string_t)(uint32_t, uint32_t, uint32_t);
+  static roctracer_op_string_t func = nullptr;
+  Dispatch<ExternLibRoctracer>::init(ExternLibRoctracer::name,
+                                     &ExternLibRoctracer::lib);
+  if (func == nullptr)
+    func = reinterpret_cast<roctracer_op_string_t>(
+        dlsym(ExternLibRoctracer::lib, "roctracer_op_string"));
```

Comment:
Does it the API's name or the kernel's name?

Reply:
The API name (for ACTIVITY_DOMAIN_HIP_API), e.g. hipLaunchKernel and the optype (for ACTIVITY_DOMAIN_HIP_OPS), e.g. KernelExecution, MemCopyHtoD.  Human readable description of what the thing is.

Kernel names are in a union in the activity record for the async ops.  You can find them in the api callback (when they are being launched) as well.

Reply:
Got it. 

I don't think we need API names. CUPTI can get "kernel name" in both callbacks and activities. I'll clean this up later.